apsuite/commisslib/measure_bbb_data.py

```diff
@@ -523,3 +523,250 @@ def waterfall_plot(axis, xs, zs, data):
         axis.set_ylim3d(xs.min(), xs.max())
         axis.set_xlim3d(zs.min(), zs.max())
         axis.set_zlim3d(0, data.max())
+
+
+class TuneShiftParams(_ParamsBaseClass):
+    """."""
+
+    TIME_REV = 864 / 499666000  # s
+    WAIT_INJ = 0.2  # s
+    DEF_TOL_CURRENT = 0.01  # mA
+
+    def __init__(self):
+        """."""
+        super().__init__()
+        self.plane = 'H'  # 'H' or 'V'
+        self.kickh = -25/1000  # mrad
+        self.kickv = +20/1000  # mrad
+        self.wait_bbb = 9  # s
+        self.currents = list()
+        self.filename = ''
+
+    def __str__(self):
+        """."""
+        dtmp = '{0:10s} = {1:9d}  {2:s}\n'.format
+        ftmp = '{0:10s} = {1:9.3f}  {2:s}\n'.format
+        ltmp = '{0:6.3f},'.format
+        stg = f"{'plane':10s} = {self.plane:9s}\n"
+        stg += ftmp('kickh', self.kickh, '[mrad]')
+        stg += ftmp('kickv', self.kickv, '[mrad]')
+        stg += dtmp('wait_bbb', self.wait_bbb, '[s]')
+        stg += f"{'currents':10s} = "
+        for curr in self.currents:
+            stg += ltmp(curr)
+        stg += ' mA \n'
+        stg += f"{'filename':10s} = '{self.filename:s}'\n"
+        return stg
+
+
+class MeasTuneShift(_BaseClass):
+    """."""
+
+    def __init__(self, isonline=True):
+        """."""
+        self.devices = dict()
+        self.data = dict()
+        self.params = TuneShiftParams()
+        if isonline:
+            self.devices['bbbh'] = BunchbyBunch(BunchbyBunch.DEVICES.H)
+            self.devices['bbbv'] = BunchbyBunch(BunchbyBunch.DEVICES.V)
+            self.devices['pingh'] = PowerSupplyPU(
+                    PowerSupplyPU.DEVICES.SI_INJ_DPKCKR)
+            self.devices['pingv'] = PowerSupplyPU(
+                PowerSupplyPU.DEVICES.SI_PING_V)
+            self.devices['egun'] = EGTriggerPS()
+            self.pingers = [self.devices['pingh'], self.devices['pingv']]
```

Comment:
maybe it is better to create this attribute as an empty list outside the `if isonline` scope.

Reply:
done!